ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift

```diff
@@ -294,18 +294,65 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
       self.swallowBackspaceTextChange = false
       return
     }
-    
+
     let contextBeforeInput = textDocumentProxy.documentContextBeforeInput ?? ""
     let selection = textDocumentProxy.selectedText ?? ""
     let contextAfterInput = textDocumentProxy.documentContextAfterInput ?? ""
     let context = "\(contextBeforeInput)\(selection)\(contextAfterInput)"
-    let bLength = contextBeforeInput.utf16.count
-    let sLength = selection.utf16.count
+    let bLength = contextBeforeInput.count
+    let sLength = selection.count
     setContextState(text: context, range: NSMakeRange(bLength, sLength))
     // Within the app, this is triggered after every keyboard input.
     // We should NOT call .resetContext() here for this reason.
   }
-  
+
+  // Pre-condition:  no text is selected.  As this is currently only called by `insertText`
+  // below, this condition is met.
+  func sendContextUpdate() {
+    let preCaretContext = textDocumentProxy.documentContextBeforeInput ?? ""
+    let postCaretContext = textDocumentProxy.documentContextAfterInput ?? ""
+
+    let updater = { (_ before: String, _ after: String) -> Void in
+      let contextWindowText = "\(before)\(after)"
+
+      // On the KMW app/webview side, `location` is SMP-aware.
+      // .utf16:  treats SMPs as two char.  (Tested with U+1D5BA)
+      // Without .utf16 - treats SMPs as one char.
+      let range = NSRange(location: before.count, length: 0)
+
+      os_log("setContextState(text: `%@`, range initializer location: %d, range.location: %d", log: KeymanEngineLogger.engine, type: .debug, contextWindowText, before.count, range.location)
+      self.setContextState(text: contextWindowText, range: range)
+    }
+
+    updater(preCaretContext, postCaretContext)
+
+    if preCaretContext == "" {
+      /* The `textDocumentProxy` abstraction is documented (in passing) as involving
+       * inter-process communication.  It is thus asynchronous.  Despite all attempts to prod
+       * it, the context window is only ever updated if an attempt to make an actual *edit*
+       * outside of the context window occurs.  The first such edit will NOT have
+       * available synchronous data... but a bit of an async wait will usually succeed in
+       * getting the update.
+       *
+       * 33ms seemed sufficient.  Can we go lower? (~30 Hz rate)
+       * Success with 20ms. (50 Hz rate)
+       * 1ms is not sufficient, nor is 10ms.   :(
+       *
+       * Note:  these notes were taken via Simulator, not on a physical device;
+       * there's no guarantee (yet) that the times will be the same.
+       * But something refresh-rate related is a fairly reasonable assumption.
+       */
+      DispatchQueue.main.asyncAfter(deadline: .now() + 0.33) { //0.033) {  // contrast:  held backspace - every 0.1.
```

Comment:
I can't imagine that this will be reliable -- if the device is busy for any reason then it's quite likely that we'll miss the deadline.

Also, isn't this 1/3 of a second, not 33 msec?

Reply:
> Also, isn't this 1/3 of a second, not 33 msec?

Oh, whoops.  Can fix that.

> I can't imagine that this will be reliable -- if the device is busy for any reason then it's quite likely that we'll miss the deadline.

And that's _**fine**_.  

If we miss it, we miss it and will require +1 backspace before the new context window becomes fully available.  That's the only penalty.  It would be _much_ worse if we tried to directly prod / manipulate the context, as _that_ would result in side-effects and far, far higher fragility due to the manipulations.

Reply:
ok